### PR TITLE
Add basic object accessor tests and fix a few bugs

### PR DESCRIPTION
**Pull request by tgoyne**
The object accessor code was never updated to reflect sync's requirements (no setting the PK after creating objects and default values have to be explicitly flagged for them to be merged correctly).

Quite a bit more is still needed here (this doesn't fully cover the object accessor code and even the parts that are covered aren't particularly thorough), but it's better than nothing.

### CODE REVIEW

File: src/object_accessor.hpp [hunk 1]
```diff
@@ -128,91 +94,75 @@ struct MutationOutsideTransactionException : public std::logic_error {
...
+            table.set_timestamp(column, row, Accessor::to_timestamp(ctx, value), is_default);
             break;
         case PropertyType::Object: {
-            if (Accessor::is_null(ctx, value)) {
-                m_row.nullify_link(column);
-            }
-            else {
-                m_row.set_link(column, Accessor::to_object_index(ctx, m_realm, value, property.object_type, try_update));
-            }
+            table.set_link(column, row, Accessor::to_object_index(ctx, m_realm, value, property.object_type, try_update), is_default);
```
Comment (bdash) [inline src/object_accessor.hpp]: This appears to expect `to_object_index` to return `npos` if the value is null. The test implementation of `to_object_index` doesn't appear to do this, nor does JS's existing implementation.

File: tests/object.cpp [hunk 2]
```diff
@@ -18,23 +18,99 @@
...
+    static util::Any from_string(Defaults*, StringData v) { return std::string(v); }
+    static util::Any from_timestamp(Defaults*, Timestamp v) { return v; }
+    static util::Any from_list(Defaults*, List v) { return v; }
+    static util::Any from_results(Defaults*, Results v) { return v; }
+    static util::Any from_object(Defaults*, Object v) { return v; }
+
+    static bool is_null(Defaults*, util::Any& v) { return !v.has_value(); }
+    static util::Any null_value(Defaults*) { return {}; }
+
+    static size_t to_existing_object_index(Defaults*, SharedRealm, util::Any &) { return 0; }
```
Comment (bdash) [inline tests/object.cpp]: It'd be clearer to have this be something like `REALM_ASSERT(false && "Not implemented");`, so it's obvious to anyone adding new tests that this needs to be fleshed out before it can be tested.

File: tests/object.cpp [hunk 3]
```diff
@@ -18,23 +18,99 @@
...
+    static size_t to_object_index(Defaults* d, SharedRealm realm, util::Any& value, std::string const& object_type, bool update)
+    {
+        if (auto object = any_cast<Object>(&value)) {
```
Comment (bdash) [inline tests/object.cpp]: If's with a single line body are brace-less elsewhere in this PR.

File: tests/util/any.hpp [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+////////////////////////////////////////////////////////////////////////////
+
+#include <typeinfo>
+#include <type_traits>
+#include <stdexcept>
+
+namespace realm {
+namespace util {
+
+// A naive implementation of C++17's std::Any
```
Comment (bdash) [inline tests/util/any.hpp]: `std::any`

File: src/object_accessor.hpp (hunk 1 above)
Comment (tgoyne) [inline src/object_accessor.hpp]: `value` can't be null here (there's a check for that before the switch statement).

File: tests/util/any.hpp [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+// This does not perform the small-object optimization or make any particular
+// attempt at being performant
+class Any final {
```
Comment (bdash) [inline tests/util/any.hpp]: This is a very clear implementation 😍

File: tests/object.cpp [hunk 6]
```diff
@@ -175,4 +277,219 @@ TEST_CASE("object") {
...
+        REQUIRE(row.get_int(2) == 5);
+        REQUIRE(row.get_float(3) == 6.6f);
+        REQUIRE(row.get_double(4) == 3.3);
+        REQUIRE(row.get_string(5) == "hello");
+        REQUIRE(row.get_binary(6) == BinaryData("olleh", 5));
+        REQUIRE(row.get_timestamp(7) == Timestamp(10, 20));
+    }
+
+    SECTION("create throws for missing values if there is no default") {
+        REQUIRE_THROWS(create(AnyDict{
```
Comment (bdash) [inline tests/object.cpp]: Is it worth testing that the exception that is thrown is the exception we expect?

File: src/object_accessor.hpp [hunk 7]
```diff
@@ -128,91 +94,75 @@ struct MutationOutsideTransactionException : public std::logic_error {
...
-    verify_attached();
-
-    if (!m_realm->is_in_transaction()) {
-        throw MutationOutsideTransactionException("Can only set property values within a transaction.");
-    }
-
+    auto& table = *m_row.get_table();
     size_t column = property.table_column;
+    size_t row = m_row.get_index();
     if (property.is_nullable && Accessor::is_null(ctx, value)) {
```
Comment (bdash) [inline src/object_accessor.hpp]: Should we throw if the property isn't nullable but we're passed a null value?

File: tests/util/test_file.cpp [hunk 8]
```diff
@@ -95,9 +95,10 @@ SyncTestFile::SyncTestFile(const SyncConfig& sync_config)
     schema_mode = SchemaMode::Additive;
 }
 
-SyncTestFile::SyncTestFile(SyncServer& server)
+SyncTestFile::SyncTestFile(SyncServer& server, std::string name)
```
Comment (bdash) [inline tests/util/test_file.cpp]: The changes in test_file.{cpp,hpp} don't appear to be related to this PR?

File: src/object_accessor.hpp (hunk 7 above)
Comment (tgoyne) [inline src/object_accessor.hpp]: The code currently expects `NativeAccessor` to handle checking for type mismatches, including throwing an appropriate error when `Accessor::to_foo()` is called on a null value.

File: tests/util/test_file.cpp (hunk 8 above)
Comment (tgoyne) [inline tests/util/test_file.cpp]: This is used in the test that verifies that default values are merged correctly.

File: tests/util/any.hpp [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+}
+} // namespace util
+} // namespace realm
+
+namespace std {
+inline void swap(realm::util::Any& lhs, realm::util::Any& rhs) noexcept
+{
+    lhs.swap(rhs);
+}
+} // nsmespace std
```
Comment (bdash) [inline tests/util/any.hpp]: namespace

File: tests/util/test_file.cpp (hunk 8 above)
Comment (bdash) [inline tests/util/test_file.cpp]: So it is. I think GitHub must've decided to collapse the new file of tests when I did my find in page to check for where this was used. Sorry about that.

File: src/object_accessor.hpp (hunk 7 above)
Comment (bdash) [inline src/object_accessor.hpp]: Ok, after looking over JS's implementation I now see that they do type checking as you describe.

**Comment by tgoyne** [comment]
I hit the linker failure that's occurring on CI once, but then it worked after that. I suspect the fact that the coverage-instrumented test binary is now over 2 GB is related.

**Comment by tgoyne** [comment]
Cutting the size of the tests binary from 2GB to 1.1GB has seemingly fixed the linker failures (or it just happened to work this tie).
